Sources/third_party/Experimental/Complex.swift

```diff
@@ -0,0 +1,325 @@
+                if d.isNaN { d = T(signOf: d, magnitudeOf: 0) }
+                recalculate = true
+            }
+            if recalculate {
+                return Complex(
+                real: .infinity * (a * c - b * d),
+                imaginary: .infinity * (a * d + b * c)
+                )
+            }
+        }
+        return Complex(real: x, imaginary: y)
+    }
+
+    static func *= (lhs: inout Complex, rhs: Complex) {
+        lhs = lhs * rhs
+    }
+
+    var magnitude: T {
+        var x = abs(real)
+        var y = abs(imaginary)
+        if x.isInfinite { return x }
+        if y.isInfinite { return y }
+        if x == 0 { return y }
+        if x < y { swap(&x, &y) }
+        let ratio = y / x
+        return x * (1 + ratio * ratio).squareRoot()
+    }
+}
+
+extension Complex : SignedNumeric {
+    @differentiable(vjp: _vjpNegate where T : Differentiable)
+    static prefix func - (operand: Complex) -> Complex {
+        return Complex(real: -operand.real, imaginary: -operand.imaginary)
+    }
+
+    mutating func negate() {
+        real.negate()
+        imaginary.negate()
+    }
+}
+
+extension Complex {
+    @differentiable(vjp: _vjpDivide(lhs:rhs:) where T : Differentiable)
+    static func / (lhs: Complex, rhs: Complex) -> Complex {
+        var a = lhs.real, b = lhs.imaginary, c = rhs.real, d = rhs.imaginary
+        var x: T
+        var y: T
+        if c.magnitude >= d.magnitude {
+            let ratio = d / c
+            let denominator = c + d * ratio
+            x = (a + b * ratio) / denominator
+            y = (b - a * ratio) / denominator
+        } else {
+            let ratio = c / d
+            let denominator = c * ratio + d
+            x = (a * ratio + b) / denominator
+            y = (b * ratio - a) / denominator
+        }
+        if x.isNaN && y.isNaN {
+            if c == 0 && d == 0 && (!a.isNaN || !b.isNaN) {
+                x = T(signOf: c, magnitudeOf: .infinity) * a
+                y = T(signOf: c, magnitudeOf: .infinity) * b
+            } else if (a.isInfinite || b.isInfinite) && c.isFinite && d.isFinite {
+                a = T(signOf: a, magnitudeOf: a.isInfinite ? 1 : 0)
+                b = T(signOf: b, magnitudeOf: b.isInfinite ? 1 : 0)
+                x = .infinity * (a * c + b * d)
+                y = .infinity * (b * c - a * d)
+            } else if (c.isInfinite || d.isInfinite) && a.isFinite && b.isFinite {
+                c = T(signOf: c, magnitudeOf: c.isInfinite ? 1 : 0)
+                d = T(signOf: d, magnitudeOf: d.isInfinite ? 1 : 0)
+                x = 0 * (a * c + b * d)
+                y = 0 * (b * c - a * d)
+            }
+        }
+        return Complex(real: x, imaginary: y)
+    }
+
+
+    static func /= (lhs: inout Complex, rhs: Complex) {
+        lhs = lhs / rhs
+    }
+}
+
+extension Complex {
+    @differentiable(vjp: _vjpComplexConjugate where T : Differentiable)
+    func complexConjugate() -> Complex {
```

Comment:
Should this be a function or a computed variable? I talked to @dan-zheng and I agree in that this should be a computed property.

Reply:
I think that it's debatable. There are precedents of making unary math operations be methods instead of computed properties. If you look at [`FloatingPoint`](https://developer.apple.com/documentation/swift/floatingpoint), computed properties defined in that protocol are often things that are component-like (e.g. [`significand`](https://developer.apple.com/documentation/swift/floatingpoint/1847298-significand), [`nextUp`](https://developer.apple.com/documentation/swift/floatingpoint/1848104-nextup)) or predicate-like (e.g. [`isSubnormal`](https://developer.apple.com/documentation/swift/floatingpoint/1641181-issubnormal)). Actual unary math operations like square root are actually defined as a nullary method ([`squareRoot()`](https://developer.apple.com/documentation/swift/floatingpoint/3017991-squareroot)).

Since complex-conjugation is more like an operation than a component, I think a method makes most sense.

Reply:
Oh okay that's a fair argument. With that then ya I'll leave it as a method 😄